src/Compilers/CSharp/Portable/Symbols/Source/TypeParameterConstraintClause.cs

```diff
@@ -39,6 +39,13 @@ internal enum TypeParameterConstraintKind
         NotNull = 0x80,
         Default = 0x100,
 
+        /// <summary>
+        /// <see cref="TypeParameterConstraintKind"/> mismatch is detected during merging process for partial type declarations.
+        /// </summary>
+        PartialMismatch = 0x200,
+        ValueTypeFromConstraintTypes = 0x400, // Not set if any flag from AllValueTypeKinds is set
+        ReferenceTypeFromConstraintTypes = 0x800,
```

Comment:
nit: Consider adding a similar comment: `// Not set if any flag from AllReferenceTypeKinds is set`

Reply:
> Consider adding a similar comment: // Not set if any flag from AllReferenceTypeKinds is set

That would be incorrect.

---
In reply to: [520174545](https://github.com/dotnet/roslyn/pull/49227#discussion_r520174545) [](ancestors = 520174545)

Reply:
Ah, I didn't realize that.
What's the benefit of having a different design for value vs. reference type constraints?

---
In reply to: [520178986](https://github.com/dotnet/roslyn/pull/49227#discussion_r520178986) [](ancestors = 520178986,520174545)

Reply:
> What's the benefit of having a different design for value vs. reference type constraints?

Requirements are different, design simply follows the requirements

---
In reply to: [520181841](https://github.com/dotnet/roslyn/pull/49227#discussion_r520181841) [](ancestors = 520181841,520178986,520174545)